File: parallel-programming-with-ms-dotnet/CSharp/Chapter2/CreditReview/Account.cs
//===============================================================================
// Microsoft patterns & practices
// Parallel Programming Guide
//===============================================================================
// Copyright © Microsoft Corporation.  All rights reserved.
// This code released under the terms of the 
// Microsoft patterns & practices license (http://parallelpatterns.codeplex.com/license).
//===============================================================================

using System;
using Microsoft.Practices.ParallelGuideSamples.Utilities;

namespace Microsoft.Practices.ParallelGuideSamples.CreditReview
{
    /// <summary>
    /// One customer's account data: array of monthly balances, also predictions and warnings
    /// </summary>
    class Account
    {
        public double[] Balance { get; set; }       // array of monthly balances
        public double Overdraft { get; set; }       // overdraft limit, issue warning if predicted balance is less

        public double SeqPrediction { get; set; }   // predicted future balance from sequential calculation
        public double ParPrediction { get; set; }   // from parallel
        public double PlinqPrediction { get; set; } // from PLINQ 

        public bool SeqWarning { get; set; }        // warning flag based on prediction from sequential calculation
        public bool ParWarning { get; set; }        // from parallel
        public bool PlinqWarning { get; set; }      // from PLINQ 

        /// <summary>
        /// Constructor, allocate balance history for months, assign overdraft
        /// </summary>
        public Account(int months, double overdraft)
        {
            Balance = new double[months];
            Overdraft = overdraft;
        }

        /// <summary>
        /// Assign balance history to vary randomly around randomly assigned trend
        /// </summary>
        public void AssignRandomTrend(Trend goodBalance, Trend badBalance, double variation, Random random)
        {
            // choose random trend
            const double rateScale = 100.0;
            const double balanceScale = 100.0;
            double rateMean = (goodBalance.Slope + badBalance.Slope) / 2;
            double initialBalanceMean = (goodBalance.Intercept + badBalance.Intercept) / 2;
            double rate = rateMean + rateScale * random.NextDouble();
            double initialBalance = initialBalanceMean + balanceScale * random.NextDouble();
            Trend trend = new Trend { Slope = rate, Intercept = initialBalance };

            // balance history is trend plus noise
            for (int i = 0; i < Balance.Length; i++)
            {
            	Balance[i] = trend.Predict(i) + variation * random.NextDouble();
            }
        }

        /// <summary>
        /// Print balances for months starting at firstMonth
        /// </summary>
        public void PrintBalance(int firstMonth, int months)
        {
            for (int month = firstMonth; month < firstMonth + months; month++)
            {
                if (month < Balance.Length)
                {
                    Console.Write("{0,9:F}", Balance[month]);
                }
                else
                {
                    Console.Write("        "); // line up columns even if data missing
                }
            }
            // no WriteLine, may want to print more 
        }
    }
}

File: parallel-programming-with-ms-dotnet/FSharp/Chapter2/CreditReview/Account.fs
//===============================================================================
// Microsoft patterns & practices
// Parallel Programming Guide
//===============================================================================
// Copyright © Microsoft Corporation.  All rights reserved.
// This code released under the terms of the 
// Microsoft patterns & practices license (http://parallelpatterns.codeplex.com/license).
//===============================================================================

namespace Microsoft.Practices.ParallelGuideSamples.CreditReview

open System
open Microsoft.Practices.ParallelGuideSamples.Utilities

/// One customer's account data: array of monthly balances, also predictions and warnings
type Account(months, overdraft:float) = 
    // Constructor, allocate balance history for months, assign overdraft
    let balance = Array.zeroCreate months

    let mutable seqPred = 0.0
    let mutable parPred = 0.0
    let mutable plinqPred = 0.0
    let mutable pseqPred = 0.0

    let mutable seqWarn = false
    let mutable parWarn = false
    let mutable plinqWarn = false
    let mutable pseqWarn = false

    static let rateScale = 100.0
    static let balanceScale = 100.0

    member x.Overdraft = overdraft
    member x.Balance = balance

    // Future balance predicted by sequential calculation
    member x.SeqPrediction 
        with get() = seqPred and set(v) = seqPred <- v
    // Future balance predicted by parallel calculation
    member x.ParPrediction 
        with get() = parPred and set(v) = parPred <- v
    // Future balance predicted by PLINQ calculation
    member x.PlinqPrediction 
        with get() = plinqPred and set(v) = plinqPred <- v
    // Future balance predicted by PSeq module calculation
    member x.PSeqPrediction 
        with get() = pseqPred and set(v) = pseqPred <- v

    // Warning flag based on prediction by sequential calculation
    member x.SeqWarning 
        with get() = seqWarn and set(v) = seqWarn <- v
    // Warning flag based on prediction by parallel calculation
    member x.ParWarning 
        with get() = parWarn and set(v) = parWarn <- v
    // Warning flag based on prediction by PLINQ calculation
    member x.PlinqWarning 
        with get() = plinqWarn and set(v) = plinqWarn <- v
    // Warning flag based on prediction by PSeq module calculation
    member x.PSeqWarning 
        with get() = pseqWarn and set(v) = pseqWarn <- v


    /// Assign balance history to vary randomly around randomly assigned trend
    member x.AssignRandomTrend(goodBalance:Trend, badBalance:Trend, variation, random:Random) =
        // Choose random trend
        let rateMean = (goodBalance.Slope + badBalance.Slope) / 2.0
        let initialBalanceMean = (goodBalance.Intercept + badBalance.Intercept) / 2.0
        let rate = rateMean + rateScale * random.NextDouble()
        let initialBalance = initialBalanceMean + balanceScale * random.NextDouble()
        let trend = new Trend(slope = rate, intercept = initialBalance)

        // Balance history is trend plus noise
        for i in 0 .. balance.Length - 1 do
            balance.[i] <- trend.Predict(float i) + variation * random.NextDouble()


    /// Print balances for months starting at firstMonth
    member x.PrintBalance(firstMonth, months) =
      for month in firstMonth .. firstMonth + months - 1 do
          if month < balance.Length then
              Console.Write("{0,9:F}", balance.[month]);
          else
              Console.Write("        "); // line up columns even if data missing
